### PR TITLE
python3Packages.datafusion: init at 0.4.0

### PR DESCRIPTION
**Pull request by cpcloud**
###### Motivation for this change

Add `datafusion` to nixpkgs

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cpcloud** [comment]
@GrahamcOfBorg build python3Packages.datafusion

**Comment by cpcloud** [comment]
@GrahamcOfBorg build arrow-cpp

**Comment by cpcloud** [comment]
@GrahamcOfBorg build arrow-cpp

File: pkgs/development/python-modules/datafusion/default.nix [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    sha256 = "sha256-JGyDxpfBXzduJaMF1sbmRm7KJajHYdVSj+WbiSETiY0=";
+    patches = [ ./Cargo.lock.patch ];
+  };
+
+  nativeBuildInputs = with rustPlatform; [ cargoSetupHook maturinBuildHook ];
+
+  buildInputs = lib.optionals stdenv.isDarwin [ libiconv ];
+  propagatedBuildInputs = [ numpy pandas pyarrow ];
+
+  checkInputs = [ pytest ];
```
Comment (veprbl) [inline pkgs/development/python-modules/datafusion/default.nix:63]:
```suggestion
  checkInputs = [ pytestCheckHook ];
  pytestFlagsArray = [ "--pyargs" pname ];
```

File: pkgs/development/python-modules/datafusion/default.nix [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  pythonImportsCheck = [ "datafusion" ];
+
+  checkPhase = ''
+    runHook preCheck
+    pytest --pyargs "${pname}"
+    runHook postCheck
+  '';
```
Comment (veprbl) [inline pkgs/development/python-modules/datafusion/default.nix:70]:
```suggestion
```

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 3]
```diff
@@ -213,6 +214,12 @@ stdenv.mkDerivation rec {
       ];
     in
     lib.optionalString doInstallCheck "-${builtins.concatStringsSep ":" filteredTests}";
+
+  preInstallCheck = ''
+    export HOME
+    HOME="$(mktemp -d)"
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix:220]: Why?

File: pkgs/development/python-modules/datafusion/default.nix (hunk 1 above)
Comment (cpcloud) [reply pkgs/development/python-modules/datafusion/default.nix:63]: I already tried this, it doesn't work unfortunately.

File: pkgs/development/python-modules/datafusion/default.nix (hunk 2 above)
Comment (cpcloud) [reply pkgs/development/python-modules/datafusion/default.nix:70]: See above. This doesn't work.

File: pkgs/development/libraries/arrow-cpp/default.nix (hunk 3 above)
Comment (cpcloud) [reply pkgs/development/libraries/arrow-cpp/default.nix:220]: I'll remove this.

**Comment by cpcloud** [comment]
@GrahamcOfBorg build arrow-cpp python3Packages.datafusion

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 4]
```diff
@@ -199,6 +199,7 @@ stdenv.mkDerivation rec {
         "S3RegionResolutionTest.PublicBucket"
         "S3RegionResolutionTest.RestrictedBucket"
         "TestMinioServer.Connect"
+        "TestS3FS.DeleteDir"
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix:202]: Strangely, I get a test failure with this:
```
[==========] Running 20 tests from 3 test suites.
[----------] Global test environment set-up.
[----------] 2 tests from S3OptionsTest
[ RUN      ] S3OptionsTest.FromAccessKey
[       OK ] S3OptionsTest.FromAccessKey (0 ms)
[ RUN      ] S3OptionsTest.FromAssumeRole
[       OK ] S3OptionsTest.FromAssumeRole (2 ms)
[----------] 2 tests from S3OptionsTest (3 ms total)

[----------] 2 tests from S3FileSystemRegionTest
[ RUN      ] S3FileSystemRegionTest.Default
[       OK ] S3FileSystemRegionTest.Default (2 ms)
[ RUN      ] S3FileSystemRegionTest.EnvironmentVariable
[       OK ] S3FileSystemRegionTest.EnvironmentVariable (0 ms)
[----------] 2 tests from S3FileSystemRegionTest (2 ms total)

[----------] 16 tests from TestS3FS
[ RUN      ] TestS3FS.GetFileInfoBucket
-- Arrow Fatal Error --
IOError: bind() failed: Operation not permitted
/private/tmp/nix-build-arrow-cpp-6.0.1.drv-0/apache-arrow-6.0.1/cpp/build/src/arrow/filesystem
```
This is confusing – I believe all tests used to run in darwin sandbox before

File: pkgs/development/python-modules/datafusion/default.nix (hunk 1 above)
Comment (cpcloud) [reply pkgs/development/python-modules/datafusion/default.nix:71]: It seems like `--pyargs` doesn't get passed in when structured this way, not sure why.

File: pkgs/development/libraries/arrow-cpp/default.nix (hunk 4 above)
Comment (cpcloud) [reply pkgs/development/libraries/arrow-cpp/default.nix:202]: I wonder if they should've been passing at all, since these tests access the network IIRC.

**Comment by cpcloud** [comment]
@veprbl Does this need any additional changes?

File: pkgs/development/python-modules/datafusion/default.nix [hunk 5]
```diff
@@ -0,0 +1,90 @@
+{ lib
+, stdenv
+, fetchurl
+, buildPythonPackage
+, fetchPypi
+, fetchFromGitHub
```
Comment (veprbl) [inline pkgs/development/python-modules/datafusion/default.nix:6]:
```suggestion
```

File: pkgs/development/python-modules/datafusion/default.nix [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+, rustPlatform
+, maturin
+, pytestCheckHook
```
Comment (veprbl) [inline pkgs/development/python-modules/datafusion/default.nix:9]:
```suggestion
```

**Review by veprbl — APPROVED** [review]
Diff LGTM

**Comment by veprbl** [comment]
@GrahamcOfBorg build python3Packages.datafusion

**Comment by veprbl** [comment]
This hangs up on tests/test_sql.py for me as well

**Comment by cpcloud** [comment]
@GrahamcOfBorg build python3Packages.datafusion

**Comment by cpcloud** [comment]
@veprbl Mind reviewing this again?

**Review by veprbl — APPROVED** [review]

**Comment by cpcloud** [comment]
@veprbl Thanks for the review!

**Comment by cpcloud** [comment]
@veprbl Can this be merged?

**Comment by veprbl** [comment]
Yes, of course

**Comment by cpcloud** [comment]
Awesome thanks!

**Pull request merged**